extensions/src/main/java/org/apache/falcon/extensions/mirroring/hdfsSnapshot/HdfsSnapshotMirroringExtension.java

```diff
@@ -0,0 +1,239 @@
+        }
+        additionalProperties.put(HdfsSnapshotMirrorProperties.TARGET_NN.getName(),
+                ClusterHelper.getStorageUrl(targetCluster));
+        additionalProperties.put(HdfsSnapshotMirrorProperties.TARGET_EXEC_URL.getName(),
+                ClusterHelper.getMREndPoint(targetCluster));
+        String targetKerberosPrincipal = ClusterHelper.getPropertyValue(targetCluster, SecurityUtil.NN_PRINCIPAL);
+        if (StringUtils.isNotBlank(targetKerberosPrincipal)) {
+            additionalProperties.put(HdfsSnapshotMirrorProperties.TARGET_NN_KERBEROS_PRINCIPAL.getName(),
+                    targetKerberosPrincipal);
+        }
+
+        String targetRetentionPolicy = extensionProperties.getProperty(
+                HdfsSnapshotMirrorProperties.TARGET_SNAPSHOT_RETENTION_POLICY.getName());
+        if (StringUtils.isBlank(targetRetentionPolicy)) {
+            targetRetentionPolicy = DEFAULT_RETENTION_POLICY;
+        }
+        validateRetentionPolicy(targetRetentionPolicy);
+        additionalProperties.put(HdfsSnapshotMirrorProperties.TARGET_SNAPSHOT_RETENTION_POLICY.getName(),
+                targetRetentionPolicy);
+
+        // Add jobName and jobCluster properties.
+        String jobName = extensionProperties.getProperty(ExtensionProperties.JOB_NAME.getName());
+        if (StringUtils.isBlank(jobName)) {
+            throw new FalconException("Property "
+                    + ExtensionProperties.JOB_NAME.getName() + " cannot be null");
+        }
+        additionalProperties.put(HdfsSnapshotMirrorProperties.SNAPSHOT_JOB_NAME.getName(), jobName);
+
+        String jobClusterName = extensionProperties.getProperty(ExtensionProperties.CLUSTER_NAME.getName());
+        Cluster jobCluster = ClusterHelper.getCluster(jobClusterName);
+        if (jobCluster == null) {
+            LOG.error("Cluster entity {} not found", ExtensionProperties.CLUSTER_NAME.getName());
+            throw new FalconException("Cluster entity "
+                    + ExtensionProperties.CLUSTER_NAME.getName() + " not found");
+        }
+        additionalProperties.put(HdfsSnapshotMirrorProperties.JOB_CLUSTER.getName(), jobClusterName);
+        additionalProperties.put(HdfsSnapshotMirrorProperties.JOB_NN.getName(),
+                ClusterHelper.getStorageUrl(jobCluster));
+        additionalProperties.put(HdfsSnapshotMirrorProperties.JOB_EXEC_URL.getName(),
+                ClusterHelper.getMREndPoint(jobCluster));
+        String jobKerberosPrincipal = ClusterHelper.getPropertyValue(jobCluster, SecurityUtil.NN_PRINCIPAL);
+        if (StringUtils.isNotBlank(jobKerberosPrincipal)) {
+            additionalProperties.put(HdfsSnapshotMirrorProperties.JOB_NN_KERBEROS_PRINCIPAL.getName(),
+                    jobKerberosPrincipal);
+        }
+        return additionalProperties;
+    }
+
+    private void validateRetentionPolicy(String retentionPolicy) throws FalconException {
```

Comment:
Its a good programming practice. You can refer this http://stackoverflow.com/questions/2671496/java-when-to-use-static-methods